### PR TITLE
#35-41 feature/subjects

### LINKED ISSUE #35: (SP: 2) Create "Get Subjects" endpoint
## Description:

The objective of this task is to implement the backend functionality for retrieving subjects through a new endpoint. 
This will involve setting up the `route`, `controller`, `service` layer, and `tests`, as well as updating the `Swagger` documentation to reflect the new endpoint's capabilities. The implementation should adhere to best practices in terms of security, performance, and code readability

## Requirements:

- `Route:` 
  - Create a new route in the application to handle `GET` requests for fetching  subjects
- `Service` layer: 
  - Implement a service that will handle the business logic for retrieving  subjects from the database
- `Controller:`
   - Develop a controller function to process incoming requests, validate query parameters, and return the appropriate response to the client. The `controller` should interact with the `service` layer to fetch the required data
- `Tests`: 
   - Write tests to ensure the endpoint functions correctly under various scenarios. Tests should cover successful data retrieval, handling of invalid requests
- `Swagger Documentation:` 
   - Update the `Swagger` docs to include the new `Get Subjects` endpoint, providing clear information on the request parameters, response structure, and status codes. Ensure the documentation accurately reflects the endpoint's functionality and any query parameters it supports

`Middleware Integration:` Utilize existing middleware for authentication and role-based access control. 
Ensure that the endpoint is accessible only to authorized roles.

`Error Handling:` Implement robust error handling to manage and respond to exceptions or validation errors, ensuring that the client receives meaningful error messages

`Code Quality:` Ensure the code is well-structured, commented, and follows project coding standards. Utilize helper functions and middleware as appropriate to keep the codebase `DRY` and maintainable

### LINKED ISSUE #36: (SP: 2) Implement "GetSubjectById" endpoint
## Description

The objective of this task is to implement an endpoint that retrieves a `single subject` by its unique identifier `(ID)`. This endpoint will be crucial for enabling users to view detailed information about a subject. It involves creating the `route`, `controller` logic, `service` logic, writing `tests` for the endpoint, and documenting it in `Swagger`

## Requirements:
- `Route:` 
  - Create a `GET` route `/subjects/:id` that extracts a subject ID from the route parameter and retrieves the corresponding subject details
- `Service` layer: 
  - Develop a service function that takes a `subject ID`, validates it, checks for the existence of the subject in the database, and returns the subject details. If the subject does not exist, an appropriate error should be thrown
- `Controller:` 
   - Implement a controller function that captures the `subject ID` from the route parameters and invokes a service function to retrieve the subject data. This function must appropriately handle and respond to errors, including not found errors, with the correct `HTTP` status codes
- `Tests`: 
   - Write tests to cover successful subject retrieval, retrieval of a non-existent subject, and handling of invalid subject IDs. Ensure the tests validate the response structure and HTTP status codes
- `Swagger Documentation:` 
   - Update the Swagger documentation to include the `"GetSubjectById"` endpoint, detailing the request path, expected parameters, response codes, and a description of the response body

`Error Handling:` Implement comprehensive error handling to manage database errors, non-existent subjects and invalid ID formats, providing clear and appropriate feedback to the client

`Security and Permissions:` Verify that any necessary authentication or authorization checks are in place to ensure that subject details are only disclosed to authorized users

### LINKED ISSUE #37: (SP: 2) Implement "GetNamesByCategoryId" endpoint
## Description:

The objective of this task is to implement the backend functionality for retrieving `subjects names` by their ` category ID` through a new endpoint. 
This will involve setting up the `route`, `controller`, `service` layer, and `tests`, as well as updating the `Swagger` documentation to reflect the new endpoint's capabilities. The implementation should adhere to best practices in terms of security, performance, and code readability

## Requirements:

- `Route:` 
  - Create a GET route `/categories/:id?/subjects/names` that extracts a category ID from the route parameter and retrieves the corresponding subjects names
- `Service` layer: 
  - Implement a service that will handle the business logic for retrieving subjects names from the database
- `Controller:`
   - Develop a controller function to process incoming requests, validate query parameters, and return the appropriate response to the client. The `controller` should interact with the `service` layer to fetch the required data
- `Tests`: 
   - Write tests to ensure the endpoint functions correctly under various scenarios. Tests should cover successful data retrieval, handling of invalid requests
- `Swagger Documentation:` 
   - Update the `Swagger` docs to include the new `GetNamesByCategoryId` endpoint, providing clear information on the request parameters, response structure, and status codes. Ensure the documentation accurately reflects the endpoint's functionality and any query parameters it supports

`Middleware Integration:` Utilize existing middleware for authentication and role-based access control. 
Ensure that the endpoint is accessible `only to authorized roles`

`Error Handling:` Implement robust error handling to manage and respond to exceptions or validation errors, ensuring that the client receives meaningful error messages

`Code Quality:` Ensure the code is well-structured, commented, and follows project coding standards. Utilize helper functions and middleware as appropriate to keep the codebase `DRY` and maintainable

### LINKED ISSUE #38: (SP: 2) Implement "Add Subject" endpoint
## Description

The purpose of this task is to develop the endpoint for `creating new  subjects` in our application. This will involve establishing the endpoint `route`, creating the `controller` and `service` logic, ensuring proper request data validation, writing `tests` for the endpoint, and updating our `Swagger` documentation to reflect this new capability

## Requirements:

- `Route:` 
  - Define a new `POST` route `/subjects` for subject creation. This route should only be accessible to authenticated users with specific role - `Admins`
- `Service` layer: 
  - Develop a service function that takes the subject data from the `controller`, validates it against our model requirements, and then saves it to the database. Ensure to handle any potential database errors
- `Controller:` 
   - Implement a controller function that handles the `POST` request. This function should capture the request body and pass it to the service layer for processing. It must also handle any errors and return the appropriate `HTTP` response codes and messages
- `Tests`: 
   - Write unit and integration tests to cover the new endpoint functionality. Ensure tests validate successful  subject creation, handling of invalid data, and proper access control
- `Swagger Documentation:` 
   - Update the Swagger documentation to include the new `"Create  Subjects"` endpoint. Documentation should clearly specify the request format, required body parameters, successful response structure, and any possible error responses

`Error Handling:` Implement comprehensive error handling within the endpoint to gracefully manage and respond to errors, providing clear and useful error messages to the client

`Security and Permissions:` Ensure that the endpoint checks for user authentication and authorizes only users with the role capable of creating subjects `(Admins)`

### LINKED ISSUE #39: (SP: 2) Implement "Update Subject" endpoint
## Description

This task is dedicated to developing the endpoint for updating existing `subjects` within our application. It involves setting up the route, crafting the controller and service logic, ensuring proper validation of request data, creating tests for the endpoint's functionality, and documenting the endpoint in `Swagger`

## Requirements:
- `Route:` 
  - Define a new `PATCH` route `/subjects/:id` for updating subjects. This route should be accessible only to authenticated users with specific roles ` (Admins)`, and should allow updates to specific subject fields
- `Service` layer: 
  - Develop a service function that receives the `subject ID` and update data, validates the updates against our subject model schema and then applies the updates to the database. Error handling for database operations and validation failures should be included
- `Controller:` 
   - Implement a controller function that handles the `PATCH` request. This function should retrieve the `subject ID` from the route parameters, the update data from the request body, and then call the service layer for processing. It must handle errors and return appropriate `HTTP` response codes and messages
- `Tests`: 
   - Write tests covering the endpoint, including scenarios for successful updates, attempts to update non-existent subject, invalid update data, and `unauthorized` access attempts
- `Swagger Documentation:` 
   - Update the `Swagger` documentation to include the `"Update Subject"` endpoint, detailing the request path, body parameters, possible response codes, and descriptions of each

`Error Handling:` Ensure robust error handling within the endpoint to manage and respond to errors effectively, providing clear feedback to the client

`Security and Permissions:` Confirm that the endpoint verifies user authentication, subject can only be update by `Admin`

### LINKED ISSUE #40: (SP: 2) Implement "Delete Subject" endpoint
## Description

The goal of this task is to develop an endpoint that allows for the `deletion` of a subject from the database using its unique identifier `(ID)`. This endpoint is essential for managing the lifecycle of subjects in the application. The implementation will involve setting up the `route`, creating the `controller` and `service` logic, adding necessary validations, writing `tests` for the endpoint, and documenting it in `Swagger`

## Requirements:

- `Route:` 
  - Establish a `DELETE` route `/subjects/:id` that captures a `subject's ID` from the route parameter and initiates its deletion. This route should only be accessible to authenticated users with specific role - `Admins`
- `Service` layer: 
  - Develop a service function that takes a `subject ID`, validates it, verifies the existence of the subject, and then proceeds with the deletion. If the subject is not found, it should throw an appropriate error
- `Controller:` 
   - Implement a `controller` function that retrieves the `subject ID` from the route parameters, calls a service function to delete the subject, and handles any errors that arise, returning the correct HTTP status codes
- `Tests`: 
   - Write tests to cover the deletion of an existing subject, attempts to delete non-existent subjects, and handling of unauthorized deletion attempts. Validate that the response correctly reflects the outcome of the operation
- `Swagger Documentation:` 
   - Update the Swagger documentation to include the `"Delete Subject"` endpoint, detailing the request path, required parameters, possible response codes, and descriptions of the response bodies

`Error Handling:` Implement error handling to manage potential issues, such as database errors, subject not found errors, and `unauthorized access errors,` providing appropriate feedback to the client

`Confirm Deletion:` Optionally, before performing the deletion, confirm that the user intends to delete the subject, either through a confirmation step in the UI or an additional verification in the API (could be a header or a confirmation parameter)

### LINKED ISSUE #41: Create "Subject" model
- Should create a Subject model:
Define a SubjectSchema using Mongoose that represents subject with fields for name and appearance
Ensure the schema includes validation and handle errors from consts folder for data integrity and adheres to our predefined constants for errors and validations

- Field Specifications:

Name: Required string, must be unique, represents the name of the Subject.
StudyCategory: studyCategory - Required string, must be a valid reference to a unique category in the database. Timestamps: Automatically generated fields (createdAt and updatedAt) that track the creation and last update times of the document.

- Error Handling:
Utilize predefined error constants for validation messages to maintain consistency and clarity in error responses.

- Swagger Documentation:
Document the Subject model schema in Swagger to include all fields, types, and validation messages
Ensure the documentation is clear, concise, and includes examples where applicable to assist in understanding the model structure and API usage (should check your path for swagger ex. /api-docs) for correct docs work

### PR DESCRIPTION
**Pull request by VoitV**

### CODE REVIEW

**Review by natamrshn — APPROVED** [review]

File: src/modules/subject/subjects.routes.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+router.use(authMiddleware)
+router.param('id', idValidation)
+
+router.get('/', asyncWrapper(subjectController.subjectsFind))
+router.get('/:id', asyncWrapper(subjectController.subjectFindById))
+router.get('/categories/:id', asyncWrapper(subjectController.subjectsFindByCategoryId))
+router.post('/', asyncWrapper(subjectController.subjectCreate))
+router.put('/:id', asyncWrapper(subjectController.subjectUpdate))
+router.delete('/:id', asyncWrapper(subjectController.subjectDelete))
```
Comment (AntonRomanov999) [inline src/modules/subject/subjects.routes.js:21]: Are there any restrictions for the user role? According requirements only admins should be able create, upd or delete subjects...

Comment (VoitV) [reply src/modules/subject/subjects.routes.js:21]: yes, need to add restrict To(ADMIN)

**Review by zim89 — APPROVED** [review]

**Review by nadraharr — APPROVED** [review]

**Review by melyasanya — APPROVED** [review]

**Review by AntonRomanov999 — APPROVED** [review]

File: src/modules/subject/subjects.routes.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+const {
+  roles: { ADMIN }
+} = require('~/consts/auth')
+
+router.use(authMiddleware)
+router.param('id', idValidation)
+
+router.get('/', asyncWrapper(subjectController.subjectsFind))
+router.get('/:id', asyncWrapper(subjectController.subjectFindById))
+router.get('/categories/:id', asyncWrapper(subjectController.subjectsFindByCategoryId))
```
Comment (vlad-khrychov) [inline src/modules/subject/subjects.routes.js]: We discussed that this route should be like `/categories/:id/subjects` , not `subjects/categories/:id`.

File: src/test/integration/modules/subject/subject-service.spec.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+const { MongoMemoryServer } = require('mongodb-memory-server')
+const subjectService = require('./subject.service')
+const Subject = require('./subject.model')
+
+let mongoServer
+
+beforeAll(async () => {
+  mongoServer = await MongoMemoryServer.create()
+  const uri = mongoServer.getUri()
+  await mongoose.connect(uri, { useNewUrlParser: true, useUnifiedTopology: true })
```
Comment (vlad-khrychov) [inline src/test/integration/modules/subject/subject-service.spec.js:11]: Good approach using `mongodb-memory-server`. Consider creating a task to set up all tests with this approach. I mean updating `src/test/setup.js`.

**Review by vlad-khrychov — APPROVED** [review]

**Pull request merged**